executor/builder.go

```diff
@@ -2954,6 +3035,68 @@ func (builder *dataReaderBuilder) buildProjectionForIndexJoin(ctx context.Contex
 	return e, err
 }
 
+
+// buildRangesForIndexJoin builds kv ranges for index join when the inner plan is index scan plan.
+func buildRangesForIndexJoin(ctx sessionctx.Context, lookUpContents []*indexJoinLookUpContent,
+	ranges []*ranger.Range, keyOff2IdxOff []int, cwc *plannercore.ColWithCmpFuncManager) ([]*ranger.Range, error) {
+	retRanges := make([]*ranger.Range, 0, len(ranges)*len(lookUpContents))
+	lastPos := len(ranges[0].LowVal) - 1
+	// sc := ctx.GetSessionVars().StmtCtx
+	tmpDatumRanges := make([]*ranger.Range, 0, len(lookUpContents))
+	for _, content := range lookUpContents {
+		for _, ran := range ranges {
+			for keyOff, idxOff := range keyOff2IdxOff {
+				ran.LowVal[idxOff] = content.keys[keyOff]
+				ran.HighVal[idxOff] = content.keys[keyOff]
+			}
+		}
+		if cwc == nil {
+			// Index id is -1 means it's a common handle.
+			// var tmpKvRanges []kv.KeyRange
+			// var err error
+			// if indexID == -1 {
+			// 	tmpKvRanges, err = distsql.CommonHandleRangesToKVRanges(sc, tableID, ranges)
+			// } else {
+			// 	tmpKvRanges, err = distsql.IndexRangesToKVRanges(sc, tableID, indexID, ranges, nil)
+			// }
+			// if err != nil {
+			// 	return nil, err
+			// }
+
+			// Why []*range.Range? why not []range.Range?
+			// A deep copy is need here because the old *range.Range is overwriten
+			for _, ran := range ranges {
+				retRanges = append(retRanges, ran.Clone())
+			}
+			continue
+		}
+		nextColRanges, err := cwc.BuildRangesByRow(ctx, content.row)
+		if err != nil {
+			return nil, err
+		}
+		for _, nextColRan := range nextColRanges {
+			for _, ran := range ranges {
+				ran.LowVal[lastPos] = nextColRan.LowVal[0]
+				ran.HighVal[lastPos] = nextColRan.HighVal[0]
+				ran.LowExclude = nextColRan.LowExclude
+				ran.HighExclude = nextColRan.HighExclude
+				tmpDatumRanges = append(tmpDatumRanges, ran.Clone())
+			}
+		}
+	}
+
+	if cwc == nil {
+		// sort.Slice(kvRanges, func(i, j int) bool {
+		// 	return bytes.Compare(kvRanges[i].StartKey, kvRanges[j].StartKey) < 0
+		// })
+		// return kvRanges, nil
+		fmt.Println("================= cwc is what????", retRanges)
```

Comment:
```
create table ta(id int primary key, k int, v int, key kk(k, v));
create table tb(id int primary key, k int, v int);

insert into ta(id, k ,v) values(1, 1, 1),(2, 2, 2),(3, 3, 3),(4, 4, 4);
insert into tb(id, k, v) select 2, 2, 2;
```

```
explain select /*+ INL_JOIN(tb, ta) */ ta.* from ta join tb on ta.k = tb.k and ta.v >= tb.v; 
+------------------------------+---------+-----------+--------------------------+----------------------------------------------------------------------------------------------------------------+
| id                           | estRows | task      | access object            | operator info                                                                                                  |
+------------------------------+---------+-----------+--------------------------+----------------------------------------------------------------------------------------------------------------+
| IndexJoin_9                  | 1.25    | root      |                          | inner join, inner:IndexReader_8, outer key:test.tb.k, inner key:test.ta.k, other cond:ge(test.ta.v, test.tb.v) |
| ├─TableReader_19(Build)      | 1.00    | root      |                          | data:Selection_18                                                                                              |
| │ └─Selection_18             | 1.00    | cop[tikv] |                          | not(isnull(test.tb.k)), not(isnull(test.tb.v))                                                                 |
| │   └─TableFullScan_17       | 1.00    | cop[tikv] | table:tb                 | keep order:false, stats:pseudo                                                                                 |
| └─IndexReader_8(Probe)       | 1.25    | root      |                          | index:Selection_7                                                                                              |
|   └─Selection_7              | 1.25    | cop[tikv] |                          | not(isnull(test.ta.k)), not(isnull(test.ta.v))                                                                 |
|     └─IndexRangeScan_6       | 1.25    | cop[tikv] | table:ta, index:kk(k, v) | range: decided by [eq(test.ta.k, test.tb.k) ge(test.ta.v, test.tb.v)], keep order:false, stats:pseudo          |
+------------------------------+---------+-----------+--------------------------+----------------------------------------------------------------------------------------------------------------+
```

cwc will be `ge(test.ta.v, test.tb.v)`, it seems be suffix non-eq join condition :disappointed: